be/src/exprs/agg/retention.h

```diff
@@ -0,0 +1,125 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <cstring>
+#include <limits>
+#include <type_traits>
+
+#include "column/array_column.h"
+#include "column/binary_column.h"
+#include "column/datum.h"
+#include "column/fixed_length_column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "exprs/agg/aggregate.h"
+#include "gen_cpp/Data_types.h"
+#include "glog/logging.h"
+#include "gutil/casts.h"
+#include "runtime/mem_pool.h"
+#include "thrift/protocol/TJSONProtocol.h"
+#include "udf/udf_internal.h"
+#include "util/phmap/phmap_dump.h"
+#include "util/slice.h"
+
+namespace starrocks::vectorized {
+struct RetentionState {
+    void update(const DatumArray& datum_array) {
+        auto size = datum_array.size();
+        if (!is_initial) {
+            is_initial = true;
+            boolean_vector.resize(size);
+        }
+
+        DCHECK_EQ(boolean_vector.size(), datum_array.size());
+
+        for (int i = 0; i < size; ++i) {
+            boolean_vector[i] |= datum_array[i].is_null() ? 0 : datum_array[i].get_uint8();
+        }
+    }
+
+    void serialize(BinaryColumn* column) const {
+        size_t old_size = column->get_bytes().size();
+        size_t vector_size = boolean_vector.size();
+        size_t new_size = old_size + sizeof(size_t) + vector_size * sizeof(uint8_t);
+        column->get_bytes().resize(new_size);
+
+        auto data_ptr = column->get_bytes().data();
+        memcpy(data_ptr + old_size, &vector_size, sizeof(size_t));
+        old_size += sizeof(size_t);
+        for (int i = 0; i < vector_size; ++i) {
+            memcpy(data_ptr + old_size, boolean_vector.data() + i, sizeof(uint8_t));
+            old_size += sizeof(uint8_t);
+        }
+
+        DCHECK_EQ(old_size, new_size);
+        column->get_offset().emplace_back(new_size);
+    }
+
+    void deserialize_and_merge(uint8_t* data_ptr) {
+        size_t vector_size = *reinterpret_cast<size_t*>(data_ptr);
+        data_ptr += sizeof(size_t);
+
+        if (!is_initial) {
+            is_initial = true;
+            boolean_vector.resize(vector_size);
+        }
+
+        DCHECK_EQ(boolean_vector.size(), vector_size);
+
+        for (int i = 0; i < vector_size; ++i) {
+            boolean_vector[i] |= *reinterpret_cast<uint8_t*>(data_ptr);
+            data_ptr += sizeof(uint8_t);
+        }
+    }
+
+    void finalize_to_column(ArrayColumn* array_column) const {
+        if (is_initial) {
+            size_t size = boolean_vector.size();
+            DatumArray array;
+            array.reserve(size);
+            for (int i = 0; i < size; i++) {
+                array.emplace_back((uint8_t)(boolean_vector[i] & boolean_vector[0]));
```

Comment:
Why need & boolean_vector[0] here？

Reply:
> Why need & boolean_vector[0] here？

To the definition of our Retention, condition_n(n > 1) makes sense only if first condition is met.